### ISSUE TITLE
Discussion about PostLexer API

### ISSUE THREAD

**Issue by MegaIng** [lark-parser/lark#500]
This is a greatly simplified version of what I want to do:
```
from lark import Lark, Token

class SimplePostLex:
    def process(self, stream):
        for t in stream:
            return Token.new_borrow_position('B', t.value, t)
    @property
    def always_accept(self):
        return 'A',
p = Lark("""
start: B
A: "A"
%declare B
""", parser='lalr', postlex=SimplePostLex())
p.parse('A')
```
This does not work since the lexer does not accept `A`-Tokens, even though I told him to do so. It works if I include an `A` Token somewhere in the grammar.

**Comment by erezsh** [comment]
Ideally there would be a cleaner way to do this, but it works:
```
from lark import Lark, Token

class SimplePostLex:
    def process(self, stream):
        for t in stream:
            yield Token.new_borrow_pos('B', t.value, t)

    @property
    def always_accept(self):
        return 'A',

p = Lark("""
start: B
A: "A"
_unused: A
%declare B
""", parser='lalr', postlex=SimplePostLex(), start=["start", "_unused"])

print(p.parse('A', 'start'))
```
prints:
```
Tree(start, [Token(B, 'A')])
```
The `_unused` trick is necessary because Lark throws away terminals that aren't used by any of the rules, and using `%declare A` collides with the existing definition. I could fix the latter, but as you can see, there's a workaround anyway.

**Comment by MegaIng** [comment]
I have an idea: Have one more `%command`, maybe `%always` or `%accept` that takes a list of tokens that are always matched. This would also get rid of the requirement for `always_accept` in a PostLexer and PostLexer could be an actual function instead of a class-instance. (btw. I would be nice to have an ABC for PostLexer so you don't have to guess how the class has to look like.)

**Comment by erezsh** [comment]
> PostLexer could be an actual function instead of a class-instance

That's a decent idea.

I don't really like the `always_accepts` feature that much. It's lazy design, that I hacked in by the way. 

One alternative would be to let the postlexer to change the `ignore` list at run-time. Then it could just add `NL` whenever `paren_level > 0` and remove it when it's 0. It would be more flexible. But is there any use for it, other than for indentation? If not, than maybe it's an overkill; over-engineering.

Anyway, I'll give it some thought.

**Comment by MegaIng** [comment]
I don't like `always_accept` as a feature of the PostLexer either. That is simply not the place it belongs. (It changes the behavior of the Lexer even though it is supposed to happen afterwards). 

I am not sure if I like the idea of being able to edit the ignore list:

- First of all, it doesn't actually fix this problem. 
- Secondly, it forces the 'user' to micromanage the lexer and introduces a lot of potentially bugs for the 'user' to watch out for. (adding unnecessary things to ignore, forgetting to remove them, adding them to early/late...)

**Comment by erezsh** [comment]
> I don't like always_accept as a feature of the PostLexer either. That is simply not the place it belongs

While it's a "feature" of the lexer, there is no reason for it to exist without a postlexer. It's only the fact that the postlexer manipulates the tokens (which already requires it to micromanage and risk bugs) that makes it necessary.

**Comment by MegaIng** [comment]
Fixed in #708
